omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_postgres_validator.rb

```diff
@@ -0,0 +1,322 @@
+    # Note on not escaping username: we already connected with the same attribute,
+    # so it's safe at this point.
+    r = connection.exec("SELECT rolsuper, rolcreaterole, rolcreatedb FROM pg_authid WHERE rolname='#{cs_pg_attr['db_superuser']}';")
+    # a super user may not have createrole/createdb flags set, so check for both cases
+    unless (r[0]['rolsuper'] == 't') or (r[0]['rolcreaterole'] == 't' and r[0]['rolcreatedb'] == 't')
+      fail_with err_not_superuser
+    end
+  end
+
+  def backend_verify_postgres_version(connection)
+    # Make sure the server is a supported version.
+    r = connection.exec("SHOW server_version;")
+    v = r[0]['server_version']
+    major, minor = v.split(".")
+    # Load up our required major/minor:
+    # NOTE: our current entry in version-manifest.json is 'postgres92',effectively hardcoding the version.
+    # TODO: We will want to ensure that current postgres is always labeled as 'postgres' which will have migration
+    #       impacts, OR we create a placeholder 'postgres-version' component that exists only to capture the
+    #       actual current version.
+    manifest = JSON.parse(File.read("/opt/opscode/version-manifest.json"))
+    required_major, required_minor = manifest['software']['postgresql92']['locked_version'].split(".")
+    # that- note that our current key in chefwe want to pull in our requirement from our build-time configuration
+    # and not hard-code it here.
+    unless major == required_major and minor == required_minor
+      fail_with err_bad_postgres_version(v)
+    end
+  end
+
+  def backend_verify_empty_db(connection, name)
+    # This test is only valid on our initial run - bootstrap itself is not a sufficent check,
+    # because we may have partially bootstrapped.
+    return if previous_run
+    r = connection.exec("SELECT count(*) AS result FROM pg_database WHERE datname='name'")
+    Chef::Log.fatal(r[0])
+    fail_with err_database_exists(name) unless r[0]['result'] == '0'
+  end
+
+  def connect_as(type)
+    port = cs_pg_attr.has_key?('port') ? cs_pg_attr['port'] : node_pg_attr['port']
+    host = cs_pg_attr['vip']
+    if type == :invalid_user
+      user = 'chef_server_conn_test'
+      password = 'invalid'
+    else
+      user = cs_pg_attr['db_superuser']
+      password = cs_pg_attr['db_superuser_password']
+    end
+    # We just want this to throw an exception or not - caller knows what to do with it.
+    EcPostgres.with_connection(node, 'template1', { 'db_superuser' => user,
+                                                    'db_superuser_password' => password,
+                                                    'vip' => host,
+                                                    'port' => port}) do |conn|
+       if block_given?
+         yield(conn)
+       end
+    end
+  end
+
+  def err_cannot_change_external_flag
+<<EOM
+CSPG001: The value of postgresql['external'] must be set prior to the initial
```

Comment:
Yes. To make life easier for support. Error codes are easier to grep for than message text